### DISCUSSION TITLE
Run a command on successful transpile

### DISCUSSION THREAD

**Discussion by DMeechan** [swc-project/swc#1594] [Ideas]
It would be useful if `swc` could run a command after a successful transpile when `--watch` mode is enabled.

Something like this to re-run Node with the new files:

```bash
swc ./src -d ./dist --watch --onSuccess "node dist/index.js"
```

Similar to how `tsc-watch` has `--onSuccess`:

```bash
tsc-watch --onSuccess "node dist/index.js"
```

For now, I'm using a combination of `swc` and `chokidar`: `swc` watches for changes in `./src` and transpiles them to `./dist`. Then `chokidar` runs `node` on the updated `./dist/index.js`.

It works but I expect it's not very efficient to watch both the src and dist folders for changes when `swc` already knows what files have changed. Does anyone have a better solution?

In case anyone else needs something similar, here's what the `package.json` scripts look like:

```json
{
  "scripts": {
    "dev": "rm -f dist/index.js && npm run dev:swc & npm run dev:chokidar",
    "dev:swc": "swc ./src -d ./dist --watch --log-watch-compilation",
    "dev:chokidar": "chokidar \"dist/**/*.js\" --initial --command \"node dist/index.js\" --silent",
  }
}
```

**Comment by kdy1** [comment]
Seems like a good idea to me. But I'm not sure how other cli tools work.
So I'll look for solutions of other projects first.

I'm currently planning a rewrite of cli, and I'll add an option for it if it still seems good after looking at other projects.

**Reply by mrsimb** [reply]
Seems like a very useful feature, i don't want to add nodemon or chokidar just for this

**Comment by totaland** [comment]
Has this been implemented yet? Or do we still have to use nodemon to do this?

**Reply by Fray117** [reply]
i know this issue kinda old, but in my opinion to using nodemon is better since you might solely use that on development and not in production.

my experience from project to project that senior programmer was told me to build the whole project rather than using watch for production since it could save ton of memory and utilise tree shakeable function into precision i guess.

**disclaimer: this only my opinion for the best practice, since i think it might making some swc part heavier on solely just serving this purpose**

**Comment by ThallesP** [comment]
Hey @kdy1! If you're looking for other CLI tools, there's [tsup](https://tsup.egoist.dev/#onsuccess) (which uses esbuild under the hood) and they have a similar approach, they provide the `--onSuccess` which accepts a CLI command, like this `--onSuccess "node dist/index.js"`.

I would love to see this feature because I want to watch, build, and reload, this flag would be super useful and easy because for now, we need `nodemon` for this and an extra command.

I can even help with code if necessary.

**Comment by Iuriiiiiii** [comment]
Please implement this, It is needed!

**Comment by wh1t3h47** [comment]
Hello,

I don't see this option in swc, however I'd use the following:

```bash
#LINUX only
yarn swc -d dist **/*.ts src/**/* -w --copy-files & node --watch dist/src
```

```bash
# Windows Portable
yarn add -D concurrently
yarn concurrently 'swc -d dist src/**/* -w --copy-files' 'node --watch dist/src'
```

Please note that `--watch` is experimental in node

It's not the same as a success watch build callback, but it still works (kinda well), it's what we Brazilians call "gambiarra", aka workaround

Also node got an experimental ts compiling feature, but really I wouldn't use that, as it needs to be set up correctly to work with enums and other stuff

The only drawback of the concurrent approach is that node might start before the build finished (race condition), which would make it trigger rebuild later, I'm not sure if it would trigger multiple rebuilds for each changed file or only one rebuild

If you like the answer, please give me a follow :)

**Reply by wh1t3h47** [reply]
there is yet a better way, just using chokidar-cli

```bash
yarn chokidar 'src/**/*' -c 'swc -d dist src/ --copy-files && yarn node --watch dist/src' --initial
```

```bash
yarn add  -D chokidar-cli
```

I wasn't able to implement incremental compilation however
